IonicPortals/IonicPortals/Portal.swift

```diff
@@ -1,19 +1,87 @@
 import Foundation
+import Capacitor
 import IonicLiveUpdates
 
-@objc(Portal)
-public class Portal: NSObject {
+/// The configuration of a web application to be embedded in an iOS application
+public struct Portal {
     
-    // MARK: - Instance Properties
-    public let name: String;
-    public var initialContext: Dictionary<String, Any>?
-    public var startDir: String?
+    /// The name of the portal
+    public let name: String
     
-    public var liveUpdateConfig: LiveUpdate? = nil
+    /// The root directory of the ``Portal`` relative to root of the `Bundle`
+    public let startDir: String
     
-    public init(_ name: String, _ startDir: String?) {
+    /// Any initial state required by the web application
+    public var initialContext: [String: JSValue]
+    
+    /// The `LiveUpdate` configuration used to determine the location of updated application assets.
+    public var liveUpdateConfig: LiveUpdate? = nil {
+        didSet {
+            guard let liveUpdateConfig = liveUpdateConfig else { return }
+            try? LiveUpdateManager.shared.add(liveUpdateConfig)
+        }
+    }
+    
+    /// Creates an instance of ``Portal``
+    /// - Parameters:
+    ///   - name: The name of the portal, must be unique.
+    ///   - startDir: The starting directory of the ``Portal`` relative to the root of the ``Bundle``.
+    ///     If `nil`, the portal name is used as the starting directory. Defaults to `nil`.
+    ///   - initialContext: Any initial state rqeuired by the web application. Defaults to `[:]`.
+    ///   - liveUpdateConfig: The `LiveUpdate` configuration used to determine to location of updated application assets. Defaults to `nil`.
+    public init(name: String, startDir: String? = nil, initialContext: [String: JSValue] = [:], liveUpdateConfig: LiveUpdate? = nil) {
         self.name = name
         self.startDir = startDir ?? name
+        self.initialContext = initialContext
+        self.liveUpdateConfig = liveUpdateConfig
+        if let liveUpdateConfig = liveUpdateConfig {
+            try? LiveUpdateManager.shared.add(liveUpdateConfig)
+        }
     }
 }
 
+/// The objective-c representation of ``Portal``
+@objc public class IONPortal: NSObject {
+    internal var portal: Portal
+    
+    /// The name of the portal
+    @objc public var name: String { portal.name }
+    
+    /// The root directory of the ``Portal`` relative to root of the `Bundle`
+    @objc public var startDir: String { portal.startDir }
+    
+    /// Any initial state required by the web application
+    @objc public var initialContext: [String: Any] {
+        get { portal.initialContext }
+        set {
+            guard let context = newValue as? [String: JSValue] else { return }
```

Comment:
Does `JSValue` work here?

Reply:
It does not, which was one of the things that made me want to split these objc api apart. JSValue is a swift protocol

Reply:
@Steven0351 hmm indeed I am running into issues setting `initialContext` due to above changes:
* from:
```
public var initialContext: Dictionary<String, Any>?
```
* to:
```
public var initialContext: [String: JSValue]
```
This seems to make it impossible to set the initialContext now. Before I could pass in a dictionary of anything needed but now no matter what I pass in the `initialContext` is now always empty 😢

Reply:
I'm using this initializer:
```
public init(name: String, startDir: String? = nil, initialContext: [String: JSValue] = [:] ...
```
Using `initialContext` with any map of key/values never gets sets.
Also if trying to use the property: `portal.initialContext = <any type of value here>` seems to never set it either.

Reply:
`[String: Any]` was never actually valid. Capacitor requires it to be `[String: JSValue]` as those are the only kinds of things representable passed through the capacitor bridge.

Reply:
I'll dig into this, since anything passed here that would be valid JSON object data types should be able to be cast.

Reply:
I tried passing: `{ name: 'hello'}`

Reply:
I was able to reproduce this pretty easily myself with the most basic of types. I'll just revert back to `[String: Any]` on `Portal` itself. I had it in my head that this was being passed to Capacitor for whatever reason.

Reply:
@NathanWalker give the [coerce-nsdictionary](https://github.com/ionic-team/ionic-portals-ios/tree/coerce-nsdictionary) branch a try whenever you get a chance. I also added a test suite that validates objective-c interop so you stop finding out about these bugs before I do 😅. One thing to note is that I did _not_ revert back to `[String: Any]`, but Capacitor does provide a method for coercing NSDictionary to it's JSObject type. I went this route because it was still necessary for `IONPubSub` to work because those values _are_ sent to the Capacitor side so I opted for consistent behavior through both communication interfaces.

Reply:
@Steven0351 This is perfection 👌 Works wonderfully well now (initialContext *and* pub/sub!) You're the best thank you; having too much fun with this stuff 😄

Reply:
Excellent!